onnxscript/rewriter/custom_ops/llm_rule_sets_cuda.py

```diff
@@ -0,0 +1,495 @@
+
+    @classmethod
+    def rewrite(cls, op, x, splits):
+        return op.Rotary(x, splits, side="left", domain="ai.onnx.contrib")
+
+
+class TransposeCast1(orp.RewriteRuleAsClass):
+    """Replaces ``Cast + Transpose(. perm=[1, 0])`` by ``TransposeCast2D``."""
+
+    @classmethod
+    def pattern(cls, op, x, perm, to):
+        return op.Cast(op.Transpose(x, perm=perm), to=to)
+
+    @classmethod
+    def check(cls, context, x, perm, to) -> bool:
+        if isinstance(perm, ir.RefAttr) or isinstance(to, ir.RefAttr):
+            return False
+        if perm.value != [1, 0]:
+            return False
+        if to.value not in {onnx.TensorProto.FLOAT16, onnx.TensorProto.FLOAT}:
+            return False
+        return True
+
+    @classmethod
+    def rewrite(cls, op, x, perm, to):
+        if to.value == onnx.TensorProto.FLOAT:
+            return op.Transpose2DCastFP32(x, domain="ai.onnx.contrib")
+        return op.Transpose2DCastFP16(x, domain="ai.onnx.contrib")
+
+
+class TransposeCast2(orp.RewriteRuleAsClass):
+    """Replaces ``Transpose(. perm=[1, 0]) + Cast`` by ``TransposeCast2D``."""
+
+    @classmethod
+    def pattern(cls, op, x, perm, to):
+        return op.Transpose(op.Cast(x, to=to), perm=perm)
+
+    @classmethod
+    def check(cls, context, x, perm, to) -> bool:
+        if isinstance(perm, ir.RefAttr) or isinstance(to, ir.RefAttr):
+            return False
+        if perm.value != [1, 0]:
+            return False
+        if to.value not in {onnx.TensorProto.FLOAT16, onnx.TensorProto.FLOAT}:
+            return False
+        return True
+
+    @classmethod
+    def rewrite(cls, op, x, perm, to):
+        if to.value == onnx.TensorProto.FLOAT:
+            return op.Transpose2DCastFP32(x, domain="ai.onnx.contrib")
+        return op.Transpose2DCastFP16(x, domain="ai.onnx.contrib")
+
+
+class MulAddTranspose(orp.RewriteRuleAsClass):
+    @classmethod
+    def pattern(cls, op, x, y, z):
+        return op.Transpose(op.MulAdd(x, y, z, domain="ai.onnx.contrib"), perm=[0, 2, 1, 3])
+
+    @classmethod
+    def check(cls, context, x, y, z) -> bool:
+        return True
+
+    @classmethod
+    def rewrite(cls, op, x, y, z):
+        return op.MulAdd(x, y, z, transposeMiddle=1, domain="ai.onnx.contrib")
+
+
+class AddMulTranspose(orp.RewriteRuleAsClass):
+    @classmethod
+    def pattern(cls, op, x, y, z):
+        return op.Transpose(op.AddMul(x, y, z, domain="ai.onnx.contrib"), perm=[0, 2, 1, 3])
+
+    @classmethod
+    def check(cls, context, x, y, z) -> bool:
+        return True
+
+    @classmethod
+    def rewrite(cls, op, x, y, z):
+        return op.AddMul(x, y, z, transposeMiddle=1, domain="ai.onnx.contrib")
+
+
+def llm_rule_set_cuda() -> orp.RewriteRuleSet:
+    """Returns a set of rules fusing nodes into custom kernels.
+
+    Returns:
+        RewriteRuleSet
+    """
+    return orp.RewriteRuleSet(
+        [
+            orp.make_rewrite_rule_from_class(AddSharedInput1, True),
```

Comment:
Prefer calling by keyword for boolean args to make it more readable

Reply:
done